mcstatus/status_response.py

```diff
@@ -0,0 +1,640 @@
+        )
+
+    @staticmethod
+    def _parse_description(raw_description: Union[dict, list, str]) -> str:
+        """Parse description from raw response.
+
+        :param raw_description: Raw description.
+        :return: Parsed description.
+        """
+        if isinstance(raw_description, str):
+            return raw_description
+
+        if isinstance(raw_description, dict):
+            entries = raw_description.get("extra", [])
+            end = raw_description["text"]
+        else:
+            entries = raw_description
+            end = ""
+
+        description = ""
+
+        for entry in entries:
+            for style_key, style_val in STYLE_MAP.items():
+                if entry.get(style_key):
+                    try:
+                        if isinstance(style_val, dict):
+                            style_val = style_val[entry[style_key]]
+
+                        description += f"§{style_val}"
+                    except KeyError:
+                        pass  # ignoring these key errors strips out html color codes
+            description += entry.get("text", "")
+
+        return description + end
+
+
+@dataclass
+class NewBedrockStatusResponse(MCStatusResponse):
+    """Class for storing BedrockServer data from a status response.
+
+    :param map_name: Name of the map. Can be unset.
+    :param gamemode: Gamemode of the server. Can be hidden.
+    """
+
+    players: BedrockStatusPlayers
+    version: BedrockStatusVersion
+    map_name: Optional[str]
+    gamemode: Optional[str]
+
+    @classmethod
+    def build(cls, decoded_data: List[Any], latency: float) -> Self:
+        """Build MCStatusResponse and check is it valid.
+
+        :param decoded_data: Raw decoded response object.
+        :param latency: Latency of the request.
+        :return: `BedrockStatusResponse` object.
+        """
+
+        try:
+            map_name = decoded_data[7]
+        except IndexError:
+            map_name = None
+        try:
+            gamemode = decoded_data[8]
+        except IndexError:
+            gamemode = None
+
+        return cls(
+            players=BedrockStatusPlayers(
+                online=int(decoded_data[4]),
+                max=int(decoded_data[5]),
+            ),
+            version=BedrockStatusVersion(
+                name=decoded_data[3],
+                protocol=int(decoded_data[2]),
+                brand=decoded_data[0],
+            ),
+            motd=decoded_data[1],
+            latency=latency,
+            map_name=map_name,
+            gamemode=gamemode,
+        )
+
+
+@dataclass
+class StatusPlayers(AbstractDataclass):
+    """Class for storing players info, like current online."""
+
+    online: int
+    max: int
+
+
+@dataclass
+class JavaStatusPlayers(StatusPlayers):
+    """Class which extends a `StatusPlayers` class with list of players.
+
+    :param players: List of players. Can be empty even if online > 0.
+    """
+
+    list: List[JavaStatusPlayer]
```

Comment:
I'm not sure how I feel about naming this parameter `list`, since it's a name of a builtin class. While this isn't an issue when it's only a class parameter and not a global, I personally try to avoid naming things in python the same way as python builtins.

I don't see an issue with the old name (`sample`). In fact, it's probably more descriptive as to what it is, since list may imply that this will be a list of all players on the server, whereas `sample` immediately makes it clear that it's only some sample of the players, and may not necessarily include everyone connected.

Reply:
It is not an issue because this attribute not even used in functions (like `__init__`), because all necessary functions generated by `@dataclass` decorator.

And `sample` for me (A2 level in English) is just an alias to `example`. `list` more describe it because it is actually a list with players. When I will see `[]` instead of players, first what I will do is go to documentation (docstrings in this case), so I doesn't see any problem in this name.

Reply:
sample has pretty different meaning to list. List is just a collection of some items, while a sample is more like a peek into something, for example it's commonly used like "doctor is taking a blood sample of someone to test for some disease", so it's just some part of a whole.

This word makes much more logical sense than list, and eve though using list isn't a huge issue, there's no reason to change the terminology if the original is fine.

And yes, python is able to handle having list as parameter, even though it's a global, and dataclasses abstract this issue away since they construct init automatically, but that doesn't mean it's a good practice to call variables in that way. Not everything that's technically possible to do is a good thing to do.

Reply:
> there's no reason to change the terminology if the original is fine.

There is - new overload. I planning to write tests on them, because their logic more then just build method in `NewJava` classes.

Reply:
I mean, as annoying as that will get with backwards compatibility, ensuring compatibility being difficult isn't a good reason to make changes like these, which actually make the new code worse.

Reply:
Also, how I can know what return? Because there is only one difference between new interface and old:
In new we return empty list, if there is no info, and in old we return None if there is no info.

Reply:
For now, we'd need to take something like `_sample` and make `sample` property, after deprecation period the property will be removed and `_sample` will be renamed to `sample`

Reply:
So second breaking changes update?

Reply:
why? changes, or even removals of `_` prefixed parameters aren't something I'd classify as breaking changes, these parameters were always meant to only be used internally, and the user shouldn't rely on these too hard.

Reply:
I mean, how user will using it as new class? You can't return at one time `None` (in old) and `[]` (in new).
User can call `_sample` for new interface, but this will be renamed after 2022-08.

So how you will return `[]` or `None` in one function, without second breaking changes update?

Reply:
Well, I'm not sure how much I like it, but we should make a `_initialized_deprecated` bool flag, which will be `True` if the class was initialized with the old signature, and False if it used the new one, based on that we can use property named `sample` to return either `[]` or `None` and store the actual sample value in `_sample` parameter.

Reply:
This will be annoying, however I really would like to keep using the name `sample`, since it is the official terminology for this, and we should stick to it.

Reply:
This will work only if user will initialize class, but our code initialize class and give it to user.

Reply:
Hm, that's annoying. Maybe let's just keep the same behavior then and use `None` when the list is empty, we can worry about changing this later. While this isn't ideal, I really dislike the idea of needing to rename `sample` to `list` for no real reason, when `sample` is the official terminology.

Reply:
Renamed in 17f85b061ca24404ab4864887c2aa699523c329c.